zerver/lib/message.py

```diff
@@ -1697,3 +1697,110 @@ def update_to_dict_cache(
 
     cache_set_many(items_for_remote_cache)
     return message_ids
+
+
+def visibility_policy_for_participation(
+    sender: UserProfile,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a user
+    participates in a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    if sender.is_bot:
+        return None
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION
+    ):
+        return UserTopic.VisibilityPolicy.UNMUTED
+
+    return None
+
+
+def visibility_policy_for_send_message(
+    sender: UserProfile,
+    message: Message,
+) -> Optional[int]:
+    """
+    This function determines the visibility policy to set when a message
+    is sent to a topic, depending on the 'automatically_follow_topics_policy'
+    and 'automatically_unmute_topics_policy' settings.
+    """
+    # We prioritize 'FOLLOW' over 'UNMUTE'.
+    # We need to carefully handle the following two cases:
+    #
+    # 1. When a user starts a topic
+    #    note: starting a topic is INITIATION as well as PARTICIPATION
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |           ON_INITIATION              |     FOLLOWED
+    #         ON_INITIATION                |          ON_PARTICIPATION            |     FOLLOWED
+    #         ON_INITIATION                |            ON_INITIATION             |     FOLLOWED
+    #
+    # 2. When both the policies have the same value 'ON_PARTICIPATION'.
+    # 'automatically_follow_topics_policy' | 'automatically_unmute_topics_policy' | visibility_policy
+    #       ON_PARTICIPATION               |          ON_PARTICIPATION            |     FOLLOWED
+    visibility_policy = None
+
+    if sender.is_bot:
+        return visibility_policy
+
+    visibility_policy = visibility_policy_for_participation(sender)
+
+    if visibility_policy == UserTopic.VisibilityPolicy.FOLLOWED:
+        return visibility_policy
+
+    old_accessible_messages_in_topic = UserMessage.objects.filter(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    ).exclude(message_id=message.id)
+
+    if (
+        sender.automatically_follow_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
+        and not old_accessible_messages_in_topic.exists()
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
```

Comment:
This should be `AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION`, right?

I think we might want to just use a common set of values, and name them e.g. 
`CHANGE_USER_TOPIC_ON_INITIATION`, just to avoid the possibility of mixing this up and feeling like that fact matters.

Reply:
Ah right, `AUTOMATICALLY_UNMUTE_TOPICS_ON_INITIATION`. The tests didn't fail because internally they have the same `integer` value.

@timabbott  Sorry, I'm unsure about what you meant by:
"use a common set of values, and name them e.g. CHANGE_USER_TOPIC_ON_INITIATION"

Reply:
Oh I meant to just rename both `AUTOMATICALLY_UNMUTE_TOPICS_ON_INITIATION` and `AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION` with `CHANGE_USER_TOPIC_ON_INITIATION` -- so we'd just have one set of names for the 1/2/3 values and it'd feel very clear in the code that they're supposed to be parallel.

Reply:
Alright, this is the modified code in `zerver/models.py` (removes duplication)

```python
    # The following two settings control which topics to automatically
    # 'follow' or 'unmute in a muted stream', respectively.
    # Follow or unmute a topic automatically on:
    # - PARTICIPATION: Send a message, React to a message, Participate in a poll or Edit a TO-DO list.
    # - INITIATION: Send the first message in the topic.
    # - NEVER: Never automatically follow or unmute a topic.
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION = 1
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION = 2
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER = 3
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_CHOICES = [
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION,
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER,
    ]
    automatically_follow_topics_policy = models.PositiveSmallIntegerField(
        default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER
    )
    automatically_unmute_topics_in_muted_streams_policy = models.PositiveSmallIntegerField(
        default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER
    )
```

I have made small changes to your suggestion `CHANGE_USER_TOPIC_ON_INITIATION`
* used the term `VISIBILITY_POLICY` instead of `USER_TOPIC`
* prefixed with `AUTOMATICALLY_` for readability.

without the prefix, things like `CHANGE_USER_TOPIC_NEVER` seems to convey that "never change the visibility policy" instead of "never automatically change the visibility policy"

Reply:
OK, works for me!